extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java

```diff
@@ -176,64 +161,91 @@ public void handle(RoutingContext event) {
+                response.headers().set(HttpHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, String.valueOf(allowCredentials));
                 response.headers().set(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
             }
 
-            boolean allowCredentials = corsConfig.accessControlAllowCredentials
-                    .orElseGet(() -> corsConfig.origins.isPresent() && corsConfig.origins.get().contains(origin)
-                            && !origin.equals("*"));
-
-            response.headers().set(HttpHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, String.valueOf(allowCredentials));
-
-            final Optional<List<String>> exposedHeaders = corsConfig.exposedHeaders;
-
-            if (!isConfiguredWithWildcard(exposedHeaders)) {
-                response.headers().set(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS,
-                        String.join(",", exposedHeaders.orElse(Collections.emptyList())));
+            if (request.method().equals(HttpMethod.OPTIONS)) {
+                final String requestedMethods = request.getHeader(HttpHeaders.ACCESS_CONTROL_REQUEST_METHOD);
+                final String requestedHeaders = request.getHeader(HttpHeaders.ACCESS_CONTROL_REQUEST_HEADERS);
+                //preflight request, handle it specially
+                if (requestedHeaders != null || requestedMethods != null) {
+                    handlePreflightRequest(event, requestedHeaders, requestedMethods, origin, allowsOrigin);
+                    response.end();
+                    return;
+                }
             }
-
-            if (!allowsOrigin) {
-                LOG.debug("Origin is not allowed");
+            if (allowedHeaders != null) {
+                response.headers().add(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS, allowedHeaders);
+            }
+            if (allowedMethods != null) {
+                response.headers().add(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS, allowedMethods);
+            }
+            //we check that the actual request matches the allowed methods and headers
+            if (!isMethodAllowed(request.method())) {
+                LOG.debug("Method is not allowed");
                 response.setStatusCode(403);
-                response.setStatusMessage("CORS Rejected - Invalid origin");
+                response.setStatusMessage("CORS Rejected - Invalid method");
                 response.end();
-            } else if (request.method().equals(HttpMethod.OPTIONS) && (requestedHeaders != null || requestedMethods != null)) {
-                LOG.debug("Preflight request has completed");
-                if (corsConfig.accessControlMaxAge.isPresent()) {
-                    response.putHeader(HttpHeaders.ACCESS_CONTROL_MAX_AGE,
-                            String.valueOf(corsConfig.accessControlMaxAge.get().getSeconds()));
-                }
+                return;
+            }
+            if (!allowsOrigin) {
```

Comment:
@stuartwdouglas I think this check should be moved just above the block which sets `403` for the not allowed method, otherwise the not allowed method status message will override the not allowed origin status message which is more important I guess.